### DISCUSSION TITLE
Upper limit on interaction curves?

### DISCUSSION THREAD

**Discussion by Agent6-6-6** [robbievanleeuwen/concrete-properties#12] [Ideas]
@robbievanleeuwen, many codes ACI318 and NZS3101 have a hard upper cutoff for the axial load a section can take, which seems a little different to the squash load definition in AS3600 which is the axial capacity at zero moment. Basically the interaction curve is cutoff horizontally at this level of load as per below image. This is probably something that would need to be introduced into the interaction curve code generate as an upper limit cutoff and then return the curve at the same axial load to zero moment?

Addtionally it may be useful to also report the control points and have an option to plot these points on the interaction curves?
For example points as follows are useful for interpretation of the results:-

> A = ZERO M* AT MAX COMP
> B = MAX M* AT MAX COMP
> C = ZERO STRAIN POINT
> D = 50% STRAIN POINT
> E = BALANCED STRAIN POINT
> F = ZERO AXIAL LOAD POINT
> G = MAX TENSION POINT

![image](https://user-images.githubusercontent.com/36808164/180578816-fae20d16-61af-46ea-99e5-850a921010cf.png)

I'm intending on taking a stab at a NZS3101 design code implementation.

**Comment by robbievanleeuwen** [comment]
Hi @Agent6-6-6,

Thanks for the heads up on this, I've only ever practised in Australia so may appear to have a narrow-minded view/implementation of a few things!

> This is probably something that would need to be introduced into the interaction curve code generate as an upper limit cutoff and then return the curve at the same axial load to zero moment?

Yes this could be added to the ``moment_interaction_diagram()`` method, something like ``max_comp: float = None``, then if provided cuts off the top of the diagram.

> Addtionally it may be useful to also report the control points and have an option to plot these points on the interaction curves?

Also yes, do you think it would be a good idea to add these points on top of the ``n_points`` provided to ``moment_interaction_diagram()``? E.g. calculate 24 points between decompression and pure bending **plus** points ``C``, ``D`` and ``E``. I currently have to calculate the balanced point for AS 3600 so I could take this out of ``design_codes`` and put it into the main API along with the other two points.

The way I've calculated the balanced point for AS 3600, ``get_n_ub()``, is to find the bar furthest away from the extreme compression fibre, calculate its yield strain and then work out the neutral axis depth based on this. Does this logic sound reasonable to you?

Let me know your thoughts, if so I can look at implementing the above two items.

**Reply by Agent6-6-6** [reply]
Hi @robbievanleeuwen 

Yeah I think the max_comp method will work fine and is probably very easy to implement. The value can be worked out in the applicable DesignCode class and just passed from there. Depending on the direction you're working from (max or min load), it's either the first two or last two points (last/first one is just the max_comp and zero and requires no actual calculation)

>I've only ever practised in Australia so may appear to have a narrow-minded view/implementation of a few things!

I'm sure due to codes all doing something the same but different there will be a few more unique situations to tweak and cater for!

> Also yes, do you think it would be a good idea to add these points on top of the n_points provided to moment_interaction_diagram()? 

I was thinking they were potentially a second separate list you also ran through in the same `moment_interaction_diagram()` function, so then you can plot/annotate on the plots directly at those points if the separate list is able to be returned, like a `cp_res` variable? and if user didn't need them calculated you save on some computation time?

You might then utilise that list of control points to generate a different number of data points between the control points where the interaction diagram is more prone to sudden change.

When you have an irregular layout of reinforcement or irregular sections, things can get interesting around the balance point. With quite a crude number of points you can miss these subtleties. I know in the past when doing a similar thing in excel I've always had to add further data points between the points E and F noted my interaction curve above. Some of the changes for example in ACI318 to phi around the balance point can cause some quite localised jutty out bits or sharper transitions in that region that are hard to capture if you only have a regularly spaced points. 


One other thing I've noted in starting to generate some of the NZS3101 specific code is that the way the AS3600 and other code modules will likely be structured following from the AS3600 implementation is that it ties you into utilising only the codes official stress block formulation.

For example I might want to do a check to NZS3101 standard, but utilise the more accurate/realistic EC2 Eurocode parabolic curve or the EC2 bilinear curve, or any other curve for that matter (at least in the NZ code it says we can use any shape that agrees with the results of tests.  
![image](https://user-images.githubusercontent.com/36808164/180597322-4619e4c8-c8f9-4b7a-90b7-353e3e905738.png)

So I wonder if separating out the unique stress block factors for example so they are in a single function or class, and all the phi factor variations into a single function or class is a better long term approach, and you feed from the actual code class a stress block 'type' and that single function or class has just has all the implemented choices. Then you could mix and match anything you like, and you have a few distinct locations within the code to add the new code stress block factors, phi factors, modulus of elasticity, etc.

I kind of feel that also will make adding further codes a little easier as most of the design code class is common code related to calling the appropriate analysis functions with perhaps a few things unique to that codes way of working things out. For example for NZS3101 design phi is constant at 0.85, but other codes also share the same phi varies depending on load as AS3600. But the analysis of the section is essentially the same. 

So if you required a separate analysis method for constant phi, changing phi, etc with all the actual design code equations for this grouped in a single separate class/function it might make plugging in and expanding other international codes a bit simpler for contributors as they don't have to recreate the entire design code class as you just wrap a constant phi class into the NZS3101 class for example?

One benefit of this is historic changes to codes could be also implemented, not just the lastest stress block parameters or phi values for example. Want to work out original capacity from the 2009 AS3600 code, vs 2018 AS3600 code, no problem. Just have to add the different formulation for the factors, but the underlying analysis method is essentially identical.

```
if design_code = 'NZS3101':
    if code_ver = '2006': 
        alpha = .......

if design_code = 'AS3600'
    if code_ver = '2018': 
        alpha = .......
    if code_ver = '2009': 
        alpha = .......

etc......
```

One last thing that didn't quite gel with me naming wise was the name of the `Steel` class for conventional reinforcement. I kind of wonder if `Reinf` be a more appropriate name for this, then when you get into Post tensioning as per your roadmap you can have `Strand`. At least in this part of the world the steel in reinforced concrete is typically referred to as reinforcement by engineers.  

Anyway ignore any of the above if it doesn't fit within your vision of how you see the addition of international codes working!

But great job getting this out there, I can see it has a huge amount of potential!

**Reply by robbievanleeuwen** [reply]
Thanks for the interesting food for thought @Agent6-6-6! I've got a busy day tomorrow but I'll have a deeper look into this on Monday and see where we can take it!

**Reply by connorferster** [reply]
@Agent6-6-6 @robbievanleeuwen 

* Adding the ability to specify a cut-off for the P-M interaction diagram would be great. Canadian code has this cut-off also.
* I think having those control points as a separate list (internally) would be convenient for the reasons @Agent6-6-6 described
* I don't think I quite understand this proposal:
> I wonder if separating out the unique stress block factors for example so they are in a single function or class, and all the phi factor variations into a single function or class is a better long term approach, and you feed from the actual code class a stress block 'type' and that single function or class has just has all the implemented choices.

Currently, I can apply my constant phi factor to the compressive strength directly before providing it to the `Concrete` material class so I don't see a hurdle here. I like the way that Robbie has designed the current API :)

I like to try to balance ease-of-use vs. customization in my own libraries and I often steer closer toward ease-of-use. My concern is in making the inputs and definitions steps very cumbersome. @Agent6-6-6 could you give an example what it might look like to separate out unique stress block factors and phi factors? Is this something that can be done in the specific code implementation class itself, separate or outside of the main analysis classes?
* While I agree with @Agent6-6-6 that generally we refer to "reinforcing" as opposed to "steel", I like that the class is called `Steel` because it is short to type and is a complete word, which I imagine is easier for non-English speakers to translate than an abbreviation such as `Reinf`. That being said, the way that Robbie has created the ability to define one's own stress-strain profile for the material, one could imagine someone creating a glass or carbon fiber reinforcing material. In this case, the more general term of `Reinforcing` could be more appropriate. What I also like about `Steel` is that it is a name intended to describe a material, not what the material's purpose is. It's just that in a reinforced concrete section, the reinforcing does not necessarily have to be steel. Perhaps the `Reinforcing` class could be sub-classed to create `Steel` which would allow the API to stay the same and also allow for another reinforcing type to be created based on `Reinforcing`.

**Reply by Agent6-6-6** [reply]
@connorferster, to explain....

maybe I don't understand how it's working yet as only just dipped my toes into the design code class. But in simple terms taking the AS3600 design module, the only option for phi, alpha, gamma factors are the ones documented in the AS3600 class and hardcoded to use the rectangular stress block approach right? There is no ability to tap into other stress block formulations or alternative shapes unless you completely write another AS3600_but_use_EC2_bilinear_stress_block Class?

So I guess what I was getting at in relatively simple pseudocode terms was something like the following arrangement to allow for any stress block formulation to be feed into the design code implementation using that codes phi values, max strain, etc



```
Class (AS3600)...

def init (self, max_strain = 0.003, stress_block_for_design = 'AS3600:2018')
      self.designcode = 'AS3600'
      self.stress_block_for_design = stress_block_for_design           #here you could define any standards stress block approach to be applied to a AS3600 design

def create_concrete_material(self, compressive strength, density = 2400, max_strain, stress_block_for_design) 
      ......
      alpha, gamma = stress_block_factors(stress_block_for_design) # this calls the required stress block factors from another class or function so they are all interchangable
      E_conc = E_c(design_code) #all the various formulations for E for each code reside here for simplicity in maintenance, etc
     max_comp = ........
   etc etc

the analysis stuff within AS3600 like creating the interaction curve is all going to be common between different codes that use a similar approach. 

# the stress block function might look like this, but would not be under the specific design code class (i.e. Class(AS3600) in the current implementation).

def stress_block_factors(self,stress_block_for_design):
     if stress_block_for_design = 'AS3600:2018'
              alpha= blah
              gamma = blah
          
         if stress_block_for_design = 'ACI318-19'
              alpha= blah
              gamma = blah

etc etc
```
ANyway just a thought towards how it might scale when you have many codes implemented and a lot of the AS300 code for say drawing the interaction diagrams would repeated in each code class and there wouldn't necessarily be a need for that. 

If there is a way of cross referencing or overwriting other formulations for phi, etc as you infer than maybe I've misunderstood how the design classes would work. I'm just going off the AS3600 one and how that is currently implemented and thinking how do I copy this and amend to NZS3101 for example but have the ability to leverage other stress block formulations.

>  Reinforcing class could be sub-classed to create Steel

Yeah I think that is a good way to cover future materials if the naming convention was more intended for Steel reinforcement. I guess I was mainly thinking about conventional reinforcement as the analysis methods are built around that rather than methods for say fibreglass reinforcement where the assumption are quite different (like you want the concrete to fail first before the bars fracture usually). But obviously if the roadmap is to expand to all sorts of alternative materials then the `Reinforcing>Steel` or `Reinforcing>StainlessSteel` or `Reinforcing>FRP` subclass idea makes sense to me to bake in some default stress/strain curves for those materials which ar eused as reinforcing.

**Reply by robbievanleeuwen** [reply]
Some good discussion here! Sorry I don't have time to get into it today! Just thought I'd leave [this](https://robbievanleeuwen.github.io/concrete-properties/notebooks/design_codes.html) example here, towards the bottom I do what I think you're talking about @Agent6-6-6 with switching the profiles for the same code. I just access the ``concete`` variable here but we could implement a ``set_stress_strain_profile()`` method or something similar?

**Reply by Agent6-6-6** [reply]
Hey no worries, no hurry. 

Yeah that kind of does what I'm trying to get across, a few more lines of code to define a specific stress block, no worries. But it is possible at least. 

I was imagining you'd feed in the stress block requirement as an input to the AS3600 class, so it was a one liner affair when doing a specific code check. But I appreciate how you've got to build it up independently now if you want to do something 'non-default' to the AS3600 design method.

So could I do a design using say an ACI analysis method using ACI318 class when someone implemented it, but use AS3600 concrete stress block factors (not a constant but variable values based on the values in AS3600 class)? I guess that's what is not fundamentally apparent that you can do that at the moment right? You could only do that if the stress block functions were also defined by code. 

You're only demonstrating feeding in constants for alpha and gamma for example right when you define a different stress block?

**Reply by Agent6-6-6** [reply]
@robbievanleeuwen one other small thing, where you're dealing with a variable phi, when doing the interaction curve I believe fir accuracy at the point where phi changes from variable to constant you should really be injecting an additional axial load point to capture exact point of change on the interaction curve.


For example see how at about 4000kN you're not capturing the exact axial load where the phi relationship changes direction. I guess think of it as another control point if implementing that approach and ensure all those points are part of the points forming the full interaction plot. 

![Screenshot_20220724-110411_Chrome.jpg](https://user-images.githubusercontent.com/36808164/180625654-bb22a5c0-310a-4c3b-9dc4-adfe840b8a87.jpg)

**Reply by Agent6-6-6** [reply]
Just thinking about my last post, from the design code call to generate the interaction curve, should there be a means of simply feeding some additional axial load 'control points' so they are injected into the range of axial loads being solved for when generating the interaction curve?

Those code specific axial loads are simply the points at which the phi values change (in the output above 0kN, and ~3800kN or whatever it works out at back calculating from the phi relationship). That would seem like an effective way to capture any code specific control points based on axial load?

**Reply by robbievanleeuwen** [reply]
> Yeah I think the ``max_comp`` method will work fine and is probably very easy to implement

Thinking a bit more about this, you will also need to handle this in ``ultimate_bending_capacity()`` separately. A user may just want to calculate the bending capacity at a high axial load and not implementing the cut off here would give an unconservative result.

> a ``cp_res`` variable? should there be a means of simply feeding some additional axial load 'control points' so they are injected into the range of axial loads being solved for when generating the interaction curve?

I like the idea of this. I will add this feature.

> I kind of wonder if ``Reinf`` be a more appropriate name for this

What do we think about a ``Bar`` class? Future class that would be prestressed would a ``Strand`` class? Thinking about this if we change ``Steel`` to ``Bar``, there's no need for anything to change if you use a stainless steel or FRP bar, you just provide a different density and stress-strain profile and everything still works.

> But I appreciate how you've got to build it up independently now if you want to do something 'non-default' to the AS3600 design method.

I think this is how I want it to work. Make it simple and default for the typical use case, and if a user wants to throw in a different stress-strain profile, they can do by calling another function like ``set_stress_strain_profile()``. I'd rather do this than complicate the setting up of the design code object. This also keeps a nice generic structure that can be similar to other design code implementations.

> So could I do a design using say an ACI analysis method using ACI318 class when someone implemented it, but use AS3600 concrete stress block factors (not a constant but variable values based on the values in AS3600 class)?

I'm not sure I entirely understand this one? Are you asking if you can generate an AS3600 stress block automatically based off an ACI318 material? At the moment not really, easiest would be to calculate the parameters yourself. Although you could maybe do something like this (pseudo-code) assuming you use consistent units:

```python
fc = 40

aci = ACI318()
aci_conc = aci.create_concrete_material(compressive_strength=fc)

as = AS3600()
as_conc = as.create_concrete_material(compressive_strength=fc)

# set aci profile to as3600 profile
aci_conc.ultimate_stress_strain_profile = as.ultimate_stress_strain_profile
```

Is this what you're getting at or have I completely missed the point?

> you should really be injecting an additional axial load point to capture exact point of change on the interaction curve.

This should be doable, I'll look into it.

I think I covered everything there, let me know if I missed anything!

**Reply by connorferster** [reply]
> What do we think about a Bar class? Future class that would be prestressed would a Strand class? Thinking about this if we change Steel to Bar, there's no need for anything to change if you use a stainless steel or FRP bar, you just provide a different density and stress-strain profile and everything still works.

One of the special things about concreteproperties is that it can handle sections reinforced with, say, a large I-section. This steel section would still be defined with the `Steel` material. I think that, during the process of defining materials, it does not matter what the geometry of the material is, be it a reinforcing bar or other kind of section, but only the material is defining the behaviours of the material. Would a `Bar` be a special kind of geometry? Is Strand a special kind of geometry? My question: what are the essential differences between these things that needs to be captured?

I think whether a reinforcing element is a bar or any other section is irrelevant in relation to the material defining it. By keeping these separation of concerns (geometry vs. material), it is more consistent with sectionproperties where we define a geometry and assign a material to it. I think that some kind `Material` class would be appropriate to define the interface required for any new materials beyond `Steel` however, the Material class is used by sectionproperties and concreteproperties defines/requires some different attributes than sectionproperties. Perhaps there could be advantages in harmonizing the two classes?

Thoughts?

**Reply by robbievanleeuwen** [reply]
@connorferster @Agent6-6-6 I'm going to move the discussion on the material module naming & functionality to #17 to try to keep this thread focussed on interaction curves.

**Reply by Agent6-6-6** [reply]
> Thinking a bit more about this, you will also need to handle this in ultimate_bending_capacity() separately. A user may just want to calculate the bending capacity at a high axial load and not implementing the cut off here would give an unconservative result.

Yeah agree. In NZ at least, checking for a higher load than the horizontal portion of the curve would be a failure. That result would be of no interest to me as a designer and should ideally generate a valueerror and some warning that the axial load you're checking for is greater than the maximum allowable (and ideally include the maximum calculated value so user can revise section size or reinforcement or lower axial load until it works).

**Reply by robbievanleeuwen** [reply]
I've created a few issues based off this discussion, please have a look at these and make sure I've covered everything. I'll be looking to implement these changes shortly!

1. #20
2. #21

Material module issue is referred to in #17.
